src/anonymeter/evaluators/singling_out_evaluator.py

```diff
@@ -2,129 +2,168 @@
-            expression = f"{_random_operator('boolean')}{col}"
-        elif isinstance(values, pd.CategoricalDtype):
-            expression = f"{col} {_random_operator('categorical')} {val}"
-        elif is_numeric_dtype(values):
-            expression = f"{col} {_random_operator('numerical')} {val}"
-        elif isinstance(val, str):
-            expression = f"{col} {_random_operator('categorical')} '{_escape_quotes(val)}'"
+        val = rng_default.choice(values)
+
+        data_type = column_types[col]
+        op = _random_operator(data_type, rng_to_use)
+
+        if val is None:
+            # Null checks
+            if op == operator.eq:
+                e = pl.col(col).is_null()
+            else:
+                e = ~pl.col(col).is_null()
+        elif data_type == "boolean":
+            if op == operator.eq:
+                e = pl.col(col)
+            else:
+                e = ~pl.col(col)
         else:
-            expression = f"{col} {_random_operator('categorical')} '{val}'"
+            e = op(pl.col(col), val) #type: ignore[assignment] # mypy does not understand that applying "op" to a polars expression returns a polars expression
 
-        query.append(expression)
+        exprs.append(e)
 
-    return " & ".join(query)
+    return reduce(operator.and_, exprs)
 
 
-def _random_queries(df: pd.DataFrame, n_queries: int, n_cols: int) -> List[str]:
-    random_columns = [rng.choice(df.columns, size=n_cols, replace=False).tolist() for _ in range(n_queries)]
-    unique_values = {col: df[col].unique() for col in df.columns}
+def _determine_data_type_for_random_query(dtype: pl.DataType) -> str:
+    if dtype in (pl.Boolean,):
+        return "boolean"
+    elif dtype.is_numeric():
+        return "numerical"
+    elif dtype in (pl.Utf8, pl.Categorical):
+        return "categorical"
+    return "categorical"  # Fallback
 
-    queries: List[str] = [_random_query(unique_values=unique_values, cols=cols) for cols in random_columns]
 
-    return queries
+def _random_queries(
+    df: pl.DataFrame,
+    n_queries: int,
+    n_cols: int,
+    rng: Optional[np.random.Generator] = None,
+) -> List[pl.Expr]:
+    rng_to_use = rng_default if rng is None else rng
 
+    unique_values = {col: df[col].unique().to_list() for col in df.columns}
+    column_types = {
+        col: _determine_data_type_for_random_query(df[col].dtype)
+        for col in df.columns
+    }
 
-def safe_query_counts(query: str, df: pd.DataFrame) -> Optional[int]:
-    """Return number of elements satisfying a given query."""
-    try:
-        return len(df.query(query, engine="python"))
-    except Exception as ex:
-        logger.debug(f"Query {query} failed with {ex}.")
-        return None
+    queries = []
+    for _ in range(n_queries):
+        selected_cols = rng_to_use.choice(
+            df.columns, size=n_cols, replace=False
+        ).tolist()
+
+        queries.append(
+            _random_query(unique_values, selected_cols, column_types)
```

Comment:
I think you forgot to pass the rng around here (see my other comments about avoiding defaults in low-level funcitons). Also, I think named arguments are much safer in general.

```suggestion
            _random_query(
                unique_values=unique_values,
                cols=selected_cols,
                column_types=column_types,
                rng=rng_to_use)
            
```